cumulus/pallets/weight-reclaim/src/lib.rs

```diff
@@ -0,0 +1,291 @@
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Pallet and transaction extensions to reclaim PoV proof size weight after an extrinsic has been
+//! applied.
+//!
+//! This crate provides a transaction extensions and a pallet.
+//! * [`StorageWeightReclaim`] transaction extension: it must wrap the whole transaction extension
+//!   pipeline.
+//! * The pallet required for the transaction extensions weight information and benchmarks.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+use codec::{Decode, Encode};
+use cumulus_primitives_storage_weight_reclaim::get_proof_size;
+use derivative::Derivative;
+use frame_support::{
+	dispatch::{DispatchInfo, PostDispatchInfo},
+	pallet_prelude::Weight,
+};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{DispatchInfoOf, Dispatchable, PostDispatchInfoOf, TransactionExtension},
+	transaction_validity::{TransactionValidityError, ValidTransaction},
+	DispatchResult,
+};
+
+#[cfg(feature = "runtime-benchmarks")]
+pub mod benchmarks;
+#[cfg(test)]
+mod tests;
+mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+const LOG_TARGET: &'static str = "runtime::storage_reclaim_pallet";
+
+/// Pallet to use alongside the transaction extension [`StorageWeightReclaim`], the pallet provides
+/// weight information and benchmarks.
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type WeightInfo: WeightInfo;
+	}
+}
+
+/// Storage weight reclaim mechanism.
+///
+/// This extension must wrap all the transaction extensions:
+#[doc = docify::embed!("./src/tests.rs", Tx)]
+///
+/// This extension checks the size of the node-side storage proof
+/// before and after executing a given extrinsic. The difference between
+/// benchmarked and used weight is reclaimed.
+///
+/// If the benchmark was underestimating the proof size, then it is added to the block weight.
+#[derive(Encode, Decode, TypeInfo, Derivative)]
+#[derivative(
+	Clone(bound = "S: Clone"),
+	Eq(bound = "S: Eq"),
+	PartialEq(bound = "S: PartialEq"),
+	Default(bound = "S: Default")
+)]
+#[scale_info(skip_type_params(T))]
+pub struct StorageWeightReclaim<T, S>(pub S, core::marker::PhantomData<T>);
+
+impl<T, S> StorageWeightReclaim<T, S> {
+	/// Create a new `StorageWeightReclaim` instance.
+	pub fn new(s: S) -> Self {
+		Self(s, Default::default())
+	}
+}
+
+impl<T, S> From<S> for StorageWeightReclaim<T, S> {
+	fn from(s: S) -> Self {
+		Self::new(s)
+	}
+}
+
+impl<T, S: core::fmt::Debug> core::fmt::Debug for StorageWeightReclaim<T, S> {
+	fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> Result<(), core::fmt::Error> {
+		#[cfg(feature = "std")]
+		let _ = write!(f, "StorageWeightReclaim<{:?}>", self.0);
+
+		#[cfg(not(feature = "std"))]
+		let _ = f;
```

Comment:
To avoid the warning about unused variable `f`.

But I refactor for a more meaningful debug formatting in no-std in https://github.com/paritytech/polkadot-sdk/pull/6140/commits/22dc3560c2394068ffeb7ab720db90a78b8e1e0c